src/wms/src/main/java/org/geoserver/wms/featureinfo/GetFeatureInfoKvpReader.java

```diff
@@ -114,13 +114,20 @@ public Object read(Object req, Map kvp, Map rawKvp) throws Exception {
                     + "It should be a proper subset of those instead");
         }
 
-        for (MapLayerInfo l : request.getQueryLayers()) {
+        // remove no-queryable layers
+        for (MapLayerInfo l : request.getQueryLayers().toArray(new MapLayerInfo[0])) {
             LayerInfo layerInfo = l.getLayerInfo();
             if (!wms.isQueryable(layerInfo)) {
-                throw new ServiceException("Layer " + l.getName() + " is not queryable",
-                    WMSErrorCode.LAYER_NOT_QUERYABLE.get(request.getVersion()), "QUERY_LAYERS");
+                int layerIndex = getMapLayers.indexOf(l);
+                getMapLayers.remove(layerIndex);
+                request.getQueryLayers().remove(layerIndex);
```

Comment:
Query layers can be a subset of getmaplayers, the non queriable layer might not even be in the list.